### PR TITLE
Bulk detection for text entity

### PR DESCRIPTION
**Pull request by ashutoshsingh0223**
Enable bulk detection for text entities

### CODE REVIEW

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS FAILED... Please resolve issues

**Comment by ashutoshsingh0223** [comment]
Retest this please

**Comment by haptik-deployment** [comment]
UNIT TESTS FAILED... Please resolve issues

File: ner_v1/detectors/textual/text/text_detection.py [hunk 1]
```diff
@@ -281,28 +286,60 @@ def detect_entity(self, text, **kwargs):
...
+                                ' come to __city__, __city__,  i will visit __city__ next year ',
+                                ' i live in __city__ '
+                            ]
 
         Additionally this function assigns these lists to self.text_entity_values and self.original_texts attributes
         respectively.
         """
-        self._process_text(text)
+        bulk_query = False
+        if type(texts) is list:
```
Comment (srajdev) [inline ner_v1/detectors/textual/text/text_detection.py]: F821 undefined name 'texts'

File: ner_v1/detectors/textual/text/text_detection.py [hunk 2]
```diff
@@ -281,28 +286,60 @@ def detect_entity(self, text, **kwargs):
...
+            bulk_query = True
+        else:
+            texts = [texts]
```
Comment (srajdev) [inline ner_v1/detectors/textual/text/text_detection.py]: F821 undefined name 'texts'

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS FAILED... Please resolve issues

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: datastore/datastore.py [hunk 3]
```diff
@@ -236,50 +235,52 @@ def get_similar_dictionary(self, entity_name, text, fuzziness_threshold="auto:4,
...
-                                                                      search_language_script=search_language_script,
-                                                                      request_timeout=request_timeout,
-                                                                      **kwargs)
-        return results_dictionary
+            results_list = elastic_search.query.full_text_query(connection=self._client_or_connection,
+                                                                index_name=self._store_name,
+                                                                doc_type=self._connection_settings[
+                                                                    ELASTICSEARCH_DOC_TYPE],
+                                                                entity_name=entity_name,
+                                                                sentence=text,
```
Comment (chiragjn) [inline datastore/datastore.py]: - Use `texts` as argument name on Line 233. 
- Use `sentences` as argument for `elastic_search.query.full_text_query`.

Makes it clear that this function takes in list

File: datastore/elastic_search/query.py [hunk 4]
```diff
@@ -249,16 +250,16 @@ def full_text_query(connection, index_name, doc_type, entity_name, sentence, fuz
         index_name: The name of the index
         doc_type: The type of the documents that will be indexed
         entity_name: name of the entity to perform a 'term' query on
-        sentence: sentence in which entity has to be searched
+        sentence(list of strings): sentences in which entity has to be searched
```
Comment (chiragjn) [inline datastore/elastic_search/query.py]: Like mentioned above, make this `sentences` for clarity

File: datastore/elastic_search/query.py [hunk 5]
```diff
@@ -282,17 +283,27 @@ def full_text_query(connection, index_name, doc_type, entity_name, sentence, fuz
...
     """
-    data = _generate_es_search_dictionary(entity_name, sentence, fuzziness_threshold,
-                                          language_script=search_language_script)
-    kwargs = dict(kwargs, body=data, doc_type=doc_type, size=constants.ELASTICSEARCH_SEARCH_SIZE, index=index_name)
+    data = [
+        _generate_es_search_dictionary(entity_name, sentence_, fuzziness_threshold,
+                                       language_script=search_language_script) for sentence_ in sentence]
+    # kwargs = dict(kwargs, body=data, doc_type=doc_type, size=constants.ELASTICSEARCH_SEARCH_SIZE, index=index_name)
+    query_data = []
+    index_for_each_query = {}
```
Comment (chiragjn) [inline datastore/elastic_search/query.py]: This should be equal to
`{'index': index_name, 'type': doc_type}'`

File: datastore/elastic_search/query.py [hunk 6]
```diff
@@ -282,17 +283,27 @@ def full_text_query(connection, index_name, doc_type, entity_name, sentence, fuz
...
+    for query in data:
+        query_data.extend([index_for_each_query, query])
+    request = ''
+    for each in query_data:
+        request += '%s \n' % json.dumps(each)
```
Comment (chiragjn) [inline datastore/elastic_search/query.py]: For lines 286 - 296 it should be possible to  simplify logic
```python
index_ = {'index': index_name, 'type': doc_type}'
data = []
for sentence in sentences:
    query = _generate_es_search_dictionary(entity_name, sentence_, fuzziness_threshold, language_script=search_language_script)
    data.extend([json.dumps(index_), json.dumps(query)])
data = '\n'.join(data)
```

File: datastore/elastic_search/query.py [hunk 7]
```diff
@@ -301,11 +312,11 @@ def _run_es_search(connection, **kwargs):
         kwargs:
             Refer https://elasticsearch-py.readthedocs.io/en/master/api.html#elasticsearch.Elasticsearch.search
     Returns:
-        dictionary, search results from elasticsearch.ElasticSearch.search
+        dictionary, search results from elasticsearch.ElasticSearch.msearch
     """
     scroll = kwargs.pop('scroll', False)
     if not scroll:
-        return connection.search(**kwargs)
+        return connection.msearch(**kwargs)
```
Comment (chiragjn) [inline datastore/elastic_search/query.py]: Add a new argument to this function `msearch=False` only do .msearch() when `msearch` is set to True. Pass msearch=True from line 299

if `scroll` is True and `msearch` is True raise ValueError('Scrolling is not supported in msearch mode`)

File: datastore/elastic_search/query.py [hunk 8]
```diff
@@ -388,7 +399,7 @@ def _generate_es_search_dictionary(entity_name, text, fuzziness_threshold, langu
                 'should': [],
                 'minimum_should_match': 1
             }
-        }
+        }, 'from': 0, 'size': constants.ELASTICSEARCH_SEARCH_SIZE
```
Comment (chiragjn) [inline datastore/elastic_search/query.py]: remove `'from': 0` as it defaults to 0 anyway

File: ner_v1/detectors/base_detector.py [hunk 9]
```diff
@@ -158,9 +157,84 @@ def detect(self, message=None, structured_value=None, fallback_value=None, **kwa
         else:
             return None
 
+        if value and type(value[0]) is list:
```
Comment (chiragjn) [inline ner_v1/detectors/base_detector.py]: Like discussed offline this section should move to `detect_entity_bulk()`. Base class will provide a simple implementation of running everything sequentially. Derived classes can provide special implementations for `detect_entity_bulk()`. Also please use plurals in detect_entity_bulk() to make it clear that they operate with lists

File: ner_v1/api.py [hunk 10]
```diff
@@ -107,6 +148,39 @@ def text(request):
     return HttpResponse(json.dumps({'data': entity_output}), content_type='application/json')
 
 
+# def text_bulk_detection(request):
```
[the remark this answers is not in the document]
Comment (ashutoshsingh0223) [reply ner_v1/api.py]: Done

File: chatbot_ner/urls.py [hunk 11]
```diff
@@ -9,6 +9,7 @@
 
 
 urlpatterns = [
+    url(r'v1/text_bulk_detection/$', api_v1.text),
```
[the remark this answers is not in the document]
Comment (ashutoshsingh0223) [reply chatbot_ner/urls.py]: Done

File: datastore/elastic_search/query.py (hunk 8 above)
Comment (ashutoshsingh0223) [reply datastore/elastic_search/query.py]: Done

File: ner_v1/detectors/base_detector.py (hunk 9 above)
Comment (ashutoshsingh0223) [reply ner_v1/detectors/base_detector.py]: Done

File: datastore/elastic_search/query.py (hunk 7 above)
Comment (chiragjn) [reply datastore/elastic_search/query.py]: Have we done any benchmarking to be able to set `max_concurrent_searches` and `max_concurrent_shard_requests`. We should figure these out for optimal performance.
See https://www.elastic.co/guide/en/elasticsearch/reference/current/search-multi-search.html

File: ner_v1/detectors/textual/text/text_detection_model.py [hunk 12]
```diff
@@ -98,11 +121,19 @@ def detect_entity(self, text, **kwargs):
             crf_original_texts = crf_model.detect_entity(text=text)
 
         values, original_texts = super(TextModelDetector, self).detect_entity(text, **kwargs)
-
-        text_entity_verified_values, original_texts = self.combine_results(values=values,
-                                                                           original_texts=original_texts,
-                                                                           crf_original_texts=crf_original_texts)
-        self.text_entity_values, self.original_texts = text_entity_verified_values, original_texts
+        if values and type(values[0]) is list:
```
Comment (chiragjn) [inline ner_v1/detectors/textual/text/text_detection_model.py]: Revert these changes and move all this bulk implementation to `detect_entity_bulk()`. Do not call CrfDetetection in that method as we don't have a plan to add batch inference on CRF yet.
When calling `self.combine_results` in `detect_entity_bulk()`, pass `crf_original_texts=[]`

Do not assign anything to self.text_entity_values and self.original_texts, leave them untouched

File: ner_v1/detectors/textual/text/text_detection_model.py [hunk 13]
```diff
@@ -98,11 +121,19 @@ def detect_entity(self, text, **kwargs):
...
+            self.text_entity_values, self.original_texts = [], []
+            for inner_values, inner_original_texts in zip(values, original_texts):
```
Comment (chiragjn) [inline ner_v1/detectors/textual/text/text_detection_model.py]: Use six.moves.zip

File: datastore/elastic_search/query.py (hunk 5 above)
Comment (ashutoshsingh0223) [reply datastore/elastic_search/query.py]: Done

File: datastore/elastic_search/query.py (hunk 6 above)
Comment (ashutoshsingh0223) [reply datastore/elastic_search/query.py]: Done

File: datastore/elastic_search/query.py (hunk 7 above)
Comment (ashutoshsingh0223) [reply datastore/elastic_search/query.py]: Done

File: ner_v1/detectors/textual/text/text_detection_model.py (hunk 13 above)
Comment (ashutoshsingh0223) [reply ner_v1/detectors/textual/text/text_detection_model.py]: Done

File: ner_v1/detectors/textual/text/text_detection_model.py (hunk 12 above)
Comment (ashutoshsingh0223) [reply ner_v1/detectors/textual/text/text_detection_model.py]: Done

File: ner_v1/detectors/base_detector.py [hunk 14]
```diff
@@ -72,10 +72,51 @@ def _set_language_processing_script(self):
...
+            dict or None: dictionary containing entity_value, original_text and detection;
+                          entity_value is in itself a dict with its keys varying from entity to entity
+
+        Example:
+            1) Consider an example of restaurant detection from a message
+
+                messages = ['i want to order chinese from  mainland china and pizza from domminos']
+                output = detect(message=message)
+                print output
+
```
Comment (srajdev) [inline ner_v1/detectors/base_detector.py]: W293 blank line contains whitespace

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS FAILED... Please resolve issues

**Comment by haptik-deployment** [comment]
UNIT TESTS FAILED... Please resolve issues

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS FAILED... Please resolve issues

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v1/detectors/textual/text/text_detection.py [hunk 15]
```diff
@@ -154,20 +154,24 @@ def set_min_token_size_for_levenshtein(self, min_size):
         """
         self._min_token_size_for_fuzziness = min_size
 
-    def _process_text(self, text):
-        self.text = text.lower()
-        if isinstance(self.text, bytes):
-            self.text = self.text.decode('utf-8')
+    def _process_text(self, texts):
```
Comment (chiragjn) [inline ner_v1/detectors/textual/text/text_detection.py]: Move this to a module level function or staticmethod which just takes in text and lowers and decodes if bytes. This function should still operate on one text at a time and should not assign anything to self.text. Do it outside where this function is called

File: ner_v1/detectors/textual/text/text_detection.py [hunk 16]
```diff
@@ -154,20 +154,24 @@ def set_min_token_size_for_levenshtein(self, min_size):
...
         # Note: following rules have been disabled because cause problem with generating original text
-        # regx_to_process = RegexReplace([(r'[\'\/]', r''), (r'\s+', r' ')])
+        # regex_to_process = RegexReplace([(r'[\'\/]', r''), (r'\s+', r' ')])
         # self.processed_text = self.regx_to_process.text_substitute(self.processed_text)
-        self.processed_text = u' ' + self.processed_text + u' '
+        self.processed_text = [u' ' + processed_text + u' ' for processed_text in self.processed_text]
         self.tagged_text = self.processed_text
 
-    def _get_substring_from_processed_text(self, matched_tokens):
+    def _get_substring_from_processed_text(self, matched_tokens, sentence):
```
Comment (chiragjn) [inline ner_v1/detectors/textual/text/text_detection.py]: Change this to  `_get_substring_from_processed_text(self, text, matched_tokens)`
Change the internal functions (`_get_tokens_and_indices`) `text` argument to `txt`

File: ner_v1/detectors/textual/text/text_detection.py [hunk 17]
```diff
@@ -313,61 +350,68 @@ def _text_detection_with_variants(self):
...
+                    self.tagged_text[index] = _pattern.sub(self.tag, self.tagged_text[index])
+                    # Instead of dropping completely like in other entities,
+                    # we replace with tag to avoid matching non contiguous segments
+                    self.processed_text[index] = _pattern.sub(self.tag, self.processed_text[index])
+            value_final_list_.append(value_final_list)
+            original_final_list_.append(original_final_list)
+
+        return value_final_list_, original_final_list_
+
+    def _get_entity_substring_from_text(self, variant, sentence):
```
Comment (chiragjn) [inline ner_v1/detectors/textual/text/text_detection.py]: Change to `_get_entity_substring_from_text(self, text, variant)`

Comment (ashutoshsingh0223) [reply ner_v1/detectors/textual/text/text_detection.py]: Done

File: ner_v1/detectors/textual/text/text_detection.py (hunk 16 above)
Comment (ashutoshsingh0223) [reply ner_v1/detectors/textual/text/text_detection.py]: Done

File: ner_v1/detectors/textual/text/text_detection.py (hunk 15 above)
Comment (ashutoshsingh0223) [reply ner_v1/detectors/textual/text/text_detection.py]: Done

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: chatbot_ner/urls.py [hunk 18]
```diff
@@ -9,6 +9,7 @@
 
 
 urlpatterns = [
+    url(r'v1/text_bulk/$', api_v1.text),
```
Comment (chiragjn) [inline chatbot_ner/urls.py]: Start indicator missing '^'

File: datastore/elastic_search/query.py [hunk 19]
```diff
@@ -282,17 +283,23 @@ def full_text_query(connection, index_name, doc_type, entity_name, sentence, fuz
...
+    data = '\n'.join(data)
+
+    kwargs = dict(kwargs, body=data, doc_type=doc_type, index=index_name)
+    results = _run_es_search(connection, _msearch=True, **kwargs)
+    results = _parse_es_search_results(results.get("responses"))
     return results
 
 
-def _run_es_search(connection, **kwargs):
+def _run_es_search(connection, _msearch=False, **kwargs):
```
Comment (chiragjn) [inline datastore/elastic_search/query.py]: Remove _ from _msearch

File: datastore/elastic_search/query.py [hunk 20]
```diff
@@ -301,11 +308,17 @@ def _run_es_search(connection, **kwargs):
...
     """
     scroll = kwargs.pop('scroll', False)
     if not scroll:
-        return connection.search(**kwargs)
+        if _msearch is True:
+            return connection.msearch(**kwargs)
+        else:
+            return connection.search(**kwargs)
+
+    if scroll and _msearch is True:
```
Comment (chiragjn) [inline datastore/elastic_search/query.py]: can skip `is True`

File: ner_v1/chatbot/entity_detection.py [hunk 21]
```diff
@@ -221,6 +221,90 @@ def get_text(message, entity_name, structured_value, fallback_value, bot_message
...
+
+    if fuzziness:
+        fuzziness = parse_fuzziness_parameter(fuzziness)
+        text_model_detector.set_fuzziness_threshold(fuzziness)
+
+    if min_token_len_fuzziness:
+        min_token_len_fuzziness = int(min_token_len_fuzziness)
+        text_model_detector.set_min_token_size_for_levenshtein(min_size=min_token_len_fuzziness)
+
+    entity_output = text_model_detector.detect_bulk(messages=messages)
```
Comment (chiragjn) [inline ner_v1/chatbot/entity_detection.py]: call this based in `get_text` based on isinstance(message, six.string_types)

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v1/detectors/base_detector.py [hunk 22]
```diff
@@ -72,10 +72,50 @@ def _set_language_processing_script(self):
...
+            1) Consider an example of restaurant detection from a message
+
+                messages = ['i want to order chinese from  mainland china and pizza from domminos']
+                output = detect(message=message)
+                print output
+                    >> [{'detection': 'message', 'original_text': 'mainland china', 'entity_value':
+                    {'value': u'Mainland China'}}, {'detection': 'message', 'original_text': 'domminos',
+                    'entity_value': {'value': u"Domino's Pizza"}}]
+        """
+        if self._source_language_script != self._target_language_script and self._translation_enabled:
```
Comment (chiragjn) [inline ner_v1/detectors/base_detector.py]: Add a if messages is None then messages = []. Otherwise there will be Exception on line 98

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge
